library/Icingadb/Common/Macros.php

```diff
@@ -0,0 +1,97 @@
+<?php
+
+/* Icinga DB Web | (c) 2021 Icinga GmbH | GPLv2 */
+
+namespace Icinga\Module\Icingadb\Common;
+
+use Icinga\Application\Logger;
+use Icinga\Module\Icingadb\Model\Host;
+use Icinga\Module\Icingadb\Model\Service;
+
+trait Macros
+{
+    /**
+     * Known icinga macros
+     *
+     * @var string[]
+     */
+    protected $icingaMacros = [
+        'HOSTNAME'            => 'name',
+        'HOSTADDRESS'         => 'address',
+        'HOSTADDRESS6'        => 'address6',
+        'host.name'           => 'name',
+        'host.address'        => 'address',
+        'host.address6'       => 'address6',
+        'service.name'        => 'name'
+    ];
+
+    /**
+     * Get the given string with macros being resolved
+     *
+     * @param string       $input  The string in which to look for macros
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function expandMacros($input, $object)
+    {
+        if (preg_match_all('@\$([^\$\s]+)\$@', $input, $matches)) {
+            foreach ($matches[1] as $key => $value) {
+                $newValue = $this->resolveMacro($value, $object);
+                if ($newValue !== $value) {
+                    $input = str_replace($matches[0][$key], $newValue, $input);
+                }
+            }
+        }
+
+        return $input;
+    }
+
+    /**
+     * Resolve a macro based on $this->object being rendered
+     *
+     * @param string       $macro The macro to resolve
+     * @param Host|Service $object The host or service used to resolve the macros
+     *
+     * @return string
+     */
+    public function resolveMacro($macro, $object)
+    {
+        if (! array_key_exists($macro, $this->icingaMacros) && ! ($macro = $this->validateMacro($macro))) {
```

Comment:
That doesn't make sense. If the macro is unknown, why should it be validated? Why should an unknown macro only be ignored if invalid?

I don't think that entire check is required. The next one should suffice and does in the legacy implementation.

Reply:
What do you mean when you say `unknown macros`? When the macro couldn’t be found in `$icingaMacros`, it doesn't necessarily mean that the macro is unknown. I am checking here first whether the passed macro exists in this Known macros variable and if not I validate it for correctness, i.e. that it is only dot delimited and then the `validateMacro() ` function returns the actual property. And if none of these conditions is true, then I just return null, because I already logged the occurred error in `validateMacro()`.

Reply:
But why the check for the dot separation? It's not even possible to resolve chained properties with the `$obj->{$dottedString}` syntax.

Reply:
> It's not even possible to resolve chained properties with the `$obj->{$dottedString}` syntax.

That's why I'm also splitting up these chained properties here.
https://github.com/Icinga/icingadb-web/blob/0c4c8a2c210a8023a229b88c4058828b601810e8/library/Icingadb/Common/Macros.php#L89-L91

Reply:
I see. But that also matches `host.state.last_comment.author` and tries to resolve `state.last_comment.author`. The thing is, the legacy implementation didn't support this either, and I don't think we should open this up (yet).

This also requires from the user to either add `host.` or `service.` to the macro, which is of no use anyway because the implementation doesn't care that `host.name` is really only resolved on host objects. And if I can use `/wiki/page?host=$host.name$&service=$service.name$` I'd expect to get two different results. If that's not possible, we shouldn't require any dots.